### PR TITLE
Add support for atomic instruction on floating-point numbers

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
This PR adds support for atomic instruction on floating-point numbers:

* SPV_EXT_shader_atomic_float_add
* SPV_EXT_shader_atomic_float_min_max
* SPV_EXT_shader_atomic_float16_add

and fixes asm printer output for half floating-type.

### CODE REVIEW

File: llvm/test/CodeGen/SPIRV/extensions/SPV_EXT_shader_atomic_float_add/atomicrmw_faddfsub_double.ll [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+; CHECK: OpAtomicFAddEXT %[[TyFP64]] %[[DblPtr]] %[[ScopeDevice]] %[[MemSeqCst]] %[[Const42]]
+
+target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024"
+target triple = "spir64"
+
+@f = common dso_local local_unnamed_addr addrspace(1) global double 0.000000e+00, align 8
+
+define dso_local spir_func void @test1() local_unnamed_addr {
+entry:
+  %0 = atomicrmw fadd ptr addrspace(1) @f, double 42.000000e+00 seq_cst
```
Comment (michalpaszkowski) [inline llvm/test/CodeGen/SPIRV/extensions/SPV_EXT_shader_atomic_float_add/atomicrmw_faddfsub_double.ll:29]: Please name all virtual registers. While, I don't mind no names/just numbering, this has been brought up by other LLVM maintainers in periodic review of the SPIR-V backend codebase. One reason is to make editing tests easier.

File: llvm/test/CodeGen/SPIRV/extensions/SPV_EXT_shader_atomic_float_add/atomicrmw_faddfsub_double.ll [hunk 2]
```diff
@@ -0,0 +1,43 @@
+; RUN: not llc -O0 -mtriple=spirv32-unknown-unknown %s -o %t.spvt 2>&1 | FileCheck %s --check-prefix=CHECK-ERROR
+
+; RUN: llc -O0 -mtriple=spirv32-unknown-unknown --spirv-extensions=SPV_EXT_shader_atomic_float_add %s -o - | FileCheck %s
+
+; CHECK-ERROR: LLVM ERROR: The atomic float instruction requires the following SPIR-V extension: SPV_EXT_shader_atomic_float_add
```
Comment (michalpaszkowski) [inline llvm/test/CodeGen/SPIRV/extensions/SPV_EXT_shader_atomic_float_add/atomicrmw_faddfsub_double.ll:5]: Neat idea!

File: llvm/lib/Target/SPIRV/SPIRVBuiltins.td [hunk 3]
```diff
@@ -872,6 +873,44 @@ defm : DemangledGroupBuiltin<"group_non_uniform_scan_inclusive_logical_xors", Wo
...
+//
+// name is the demangled name of the given builtin.
+// opcode specifies the SPIR-V operation code of the generated instruction.
+//===----------------------------------------------------------------------===//
+class AtomicFloatingBuiltin<string name, Op operation> {
+  string Name = name;
+  Op Opcode = operation;
+}
+
+// Table gathering all the Intel sub group builtins.
```
Comment (michalpaszkowski) [inline llvm/lib/Target/SPIRV/SPIRVBuiltins.td:887]: Need to update the comment for this group of builtins

Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVBuiltins.td]: Indeed, thank you

**Comment by VyacheslavLevytskyy** [comment]
SPIR-V Tests failing isn't something connected with the PR, but an unrelated general error:

`[2277/2903] Performing download step for 'SPIRVTools'
FAILED: tools/spirv-tools/SPIRVTools-prefix/src/SPIRVTools-stamp/SPIRVTools-download /mnt/build/tools/spirv-tools/SPIRVTools-prefix/src/SPIRVTools-stamp/SPIRVTools-download 
cd /mnt/build/tools/spirv-tools/SPIRVTools-prefix/src && git clone https://github.com/KhronosGroup/SPIRV-Tools.git SPIRVTools && cd SPIRVTools && /__t/Python/3.11.8/x64/bin/python3 utils/git-sync-deps && /usr/bin/cmake -E touch /mnt/build/tools/spirv-tools/SPIRVTools-prefix/src/SPIRVTools-stamp/SPIRVTools-download
/bin/sh: 1: git: not found`

**Review by michalpaszkowski — APPROVED** [review]

**Pull request merged**